In [583]:
import numpy as np
import pandas as pd
from tqdm import tqdm
tqdm.pandas()

In [584]:
import geopandas as gpd
import geopy

In [585]:
business_raw = pd.read_csv('csv_files/yelp_academic_dataset_business.csv')
business_raw.index +=1

In [586]:
business_raw.head()

address  \
1   2818 E Camino Acequia Drive   
2          30 Eglinton Avenue W   
3  1210 8th Street SW, Unit 220   
4   6955 N Durango Dr, Ste 1116   
5             4 E University Dr   

                                          attributes             business_id  \
1                                                 {}  1SWheh84yJXfytovILXOAQ   
2  {'GoodForMeal': "{'dessert': False, 'latenight...  QXAEGFB4oINsVuTFxEYKFQ   
3  {'BusinessParking': "{'garage': False, 'street...  fcXOEZdXYeZqnQ3lGlOXmg   
4  {'BusinessParking': "{'garage': False, 'street...  f2ZWZPENViL92BrFsIgR6w   
5                                                 {}  6KgGE8B1RsR7jc9R5nuH0Q   

                                          categories         city  \
1                                  Golf, Active Life      Phoenix   
2  Specialty Food, Restaurants, Dim Sum, Imported...  Mississauga   
3  Local Services, Professional Services, Compute...      Calgary   
4                         Beauty & Spas, Hair Salons    Las Vegas   
5                American (Traditional), Restaurants        Tempe   

                                               hours  is_open   latitude  \
1                                                NaN        0  33.522143   
2  {'Monday': '9:0-0:0', 'Tuesday': '9:0-0:0', 'W...        1  43.605499   
3  {'Monday': '9:0-17:0', 'Tuesday': '9:0-17:0', ...        1  51.041771   
4  {'Monday': '8:30-20:0', 'Tuesday': '8:30-20:0'...        1  36.287312   
5                                                NaN        0  33.422192   

    longitude                         name postal_code  review_count  stars  \
1 -112.018481   Arizona Biltmore Golf Club       85016             5    3.0   
2  -79.652289   Emerald Chinese Restaurant     L5R 3E7           128    2.5   
3 -114.081109  Nucleus Information Service     T2R 1L3             5    2.0   
4 -115.289540                 Amazing Cuts       89149           116    4.5   
5 -111.939615                 Ruby Tuesday       85281             9    2.5   

  state  
1    AZ  
2    ON  
3    AB  
4    NV  
5    AZ

In [587]:
business_raw.shape

(192609, 14)

In [588]:
business_raw[business_raw["address"].isna()].shape

(7682, 14)

In [589]:
business_raw[business_raw["attributes"].isna()].shape

(28836, 14)

In [590]:
business_raw[business_raw["business_id"].isna()].shape

(0, 14)

In [591]:
business_raw[business_raw["categories"].isna()].shape

(482, 14)

In [592]:
business_raw[business_raw["city"].isna()]

address                                         attributes  \
119358     NaN  {'BusinessParking': "{'garage': False, 'street...   

                   business_id           categories city hours  is_open  \
119358  WpC53SqwoCY5AuYIFr_1eA  Soccer, Active Life  NaN   NaN        1   

         latitude   longitude                          name postal_code  \
119358  33.338097 -111.677975  NSFC Strikers Soccer Complex         NaN   

        review_count  stars state  
119358             3    2.5    AZ

In [593]:
business_raw[business_raw["hours"].isna()].shape

(44830, 14)

In [594]:
business_raw[business_raw["is_open"].isna()].shape

(0, 14)

In [595]:
business_raw[business_raw["latitude"].isna()].shape

(0, 14)

In [596]:
business_raw[business_raw["longitude"].isna()].shape

(0, 14)

In [597]:
business_raw[business_raw["name"].isna()].shape

(0, 14)

In [598]:
business_raw[business_raw["postal_code"].isna()].shape

(659, 14)

In [599]:
business_raw[business_raw["review_count"].isna()].shape

(0, 14)

In [600]:
business_raw[business_raw["stars"].isna()].shape

(0, 14)

In [601]:
business_raw[business_raw["state"].isna()].shape

(0, 14)

In [602]:
business_raw[business_raw["is_open"] == 1][business_raw["hours"].isna()]

/home/bastien/.local/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


address  \
23                            12 Rue Maguire   
34                          9801 Jane Street   
36                                       NaN   
42                           7541 N 69th Ave   
43      Holt Renfrew Centre, Concourse Level   
...                                      ...   
192579                    9480 S Eastern Ave   
192593          17019 W Greenway Rd, Ste 113   
192595                          201 N 3rd St   
192600                    8360 Charlotte Hwy   
192606                   244 Victoria Street   

                                               attributes  \
23                                                     {}   
34                                                     {}   
36                                                     {}   
42                                                    NaN   
43      {'BusinessParking': "{'garage': False, 'street...   
...                                                   ...   
192579  {'BusinessParking': "{'garage': False, 'street...   
192593  {'BusinessParking': "{'garage': False, 'street...   
192595  {'Ambience': "{'touristy': False, 'hipster': F...   
192600  {'BusinessParking': "{'garage': False, 'street...   
192606                                                 {}   

                   business_id  \
23      ILIPtZ-LVSriHVb6vMffhw   
34      FTRuQkWwQqtf9J_TTqbJtA   
36      yod_AjrQTFQs8CI_DNs4PA   
42      YJNkmzE1esc8hsQ5ANgTxA   
43      UlEPHgzjvPAD-un_m-5_fA   
...                        ...   
192579  xaF6q238K-ueSAyLBIFizA   
192593  Gfa85o6JuT9uBqi9fqC-UA   
192595  8QWPlVQ6D-OExqXoaD2Z1g   
192600  sO94GQX-F3XVekp-OxETzQ   
192606  phSaT3I_y6To1J2OdeFSUQ   

                                               categories         city hours  \
23       Juice Bars & Smoothies, Food, Cafes, Restaurants     Montréal   NaN   
34                        Restaurants, Fast Food, Burgers        Maple   NaN   
36               Photographers, Event Planning & Services        Tempe   NaN   
42                                Automotive, Auto Repair     Glendale   NaN   
43                            Hair Removal, Beauty & Spas      Toronto   NaN   
...                                                   ...          ...   ...   
192579                Women's Clothing, Fashion, Shopping    Las Vegas   NaN   
192593                 Restaurants, Fast Food, Sandwiches     Surprise   NaN   
192595  Hotels & Travel, Bar Crawl, Nightlife, Arts & ...    Las Vegas   NaN   
192600                                      Grocery, Food  Indian Land   NaN   
192606              Arts & Entertainment, Performing Arts      Toronto   NaN   

        is_open   latitude   longitude                          name  \
23            1  45.524633  -73.595568                   Grannyssima   
34            1  43.848424  -79.535712                       Wendy's   
36            1  33.336384 -111.934364      One Fine Day Photography   
42            1  33.546543 -112.207099      Rj's All Duty Automotive   
43            1  43.670369  -79.388270  Möcelle Edan Cosmetic Clinic   
...         ...        ...         ...                           ...   
192579        1  36.017147 -115.117117     Identity Fashion Boutique   
192593        1  33.623793 -112.425002                        Subway   
192595        1  36.171829 -115.142318             Vegas Pub Crawler   
192600        1  34.955435  -80.850770                        Publix   
192606        1  43.655290  -79.379541            Ed Mirvish Theatre   

       postal_code  review_count  stars state  
23         H2T 1B8             3    4.5    QC  
34         L6A 3N9             7    3.0    ON  
36           85284            10    5.0    AZ  
42           85303             4    5.0    AZ  
43         M4W 1A1             4    2.0    ON  
...            ...           ...    ...   ...  
192579       89123             3    3.5    NV  
192593       85388            39    2.5    AZ  
192595       89101            41    4.5    NV  
192600      

In [603]:
business_raw[business_raw["postal_code"].isna()][business_raw["address"].isna() == False]

/home/bastien/.local/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


address  \
480                 Gerrard Square   
560                     Market 707   
818            Plaza Hotel Main St   
4740              Sainte Catherine   
4795            100 Bloor Street W   
...                            ...   
181148               Smithfield St   
184673             Kew/Balmy Beach   
187501   Suncoast Hotel And Casino   
188370   Superstition Springs Blvd   
190048  495 36 Street NE - Safeway   

                                               attributes  \
480                                                    {}   
560     {'Ambience': "{'touristy': False, 'hipster': F...   
818                                                    {}   
4740                                                  NaN   
4795    {'BusinessParking': "{'garage': False, 'street...   
...                                                   ...   
181148                                                NaN   
184673                                                 {}   
187501  {'NoiseLevel': "u'loud'", 'Ambience': "{'roman...   
188370                                                NaN   
190048  {'BusinessParking': "{'garage': False, 'street...   

                   business_id  \
480     wTtOXvbg0cpLPtvX-JjHkA   
560     4FE9FM8uvn9CV76Bxfxogw   
818     4E3nFi3rrmKbqUjG4erKJw   
4740    OUfjqyqFPmQKX9O2HUnrKA   
4795    _OQUFcYK5zynvTnTim12Mg   
...                        ...   
181148  kNs1b6gdTg4xPiTyLcUqWQ   
184673  fLErQ66UipV6JPqtUvJOZA   
187501  2nraROl3d0-Wr9EPGkW-1A   
188370  R0gfUYIiC5MlLxGjRcYifw   
190048  bwG886o-VwYIn-Uqwezh1A   

                                               categories        city  \
480     General Dentistry, Dentists, Health & Medical,...     Toronto   
560                                Caribbean, Restaurants     Toronto   
818     Arts & Entertainment, Nightlife, Performing Ar...   Las Vegas   
4740                           Shopping Centers, Shopping    Montréal   
4795    Women's Clothing, Leather Goods, Fashion, Shop...     Toronto   
...                                                   ...         ...   
181148                                Parking, Automotive  Pittsburgh   
184673  Fitness & Instruction, Beach Equipment Rentals...     Toronto   
187501           Fast Food, Restaurants, Food, Sandwiches   Las Vegas   
188370                  Restaurants, Greek, Mediterranean        Mesa   
190048                                 Coffee & Tea, Food     Calgary   

                                                    hours  is_open   latitude  \
480                                                   NaN        1  43.670630   
560                                                   NaN        1  43.652196   
818                                                   NaN        1  36.180670   
4740                                                  NaN        1  45.481842   
4795                                                  NaN        1  43.669384   
...                                                   ...      ...        ...   
181148                                                NaN        1  40.443059   
184673                                                NaN        1  43.666745   
187501  {'Monday': '7:0-23:0', 'Tuesday': '7:0-23:0', ...        1  36.170130   
188370                                                NaN        1  33.383932   
190048                                                NaN        1  51.054088   

         longitude                                name postal_code  \
480     -79.329283                    Riverdale Dental         NaN   
560     -79.405563    T&T Roti Trinidad & Toronto Roti         NaN   
818    -115.140494                         Avant Garde         NaN   
4740    -73.592425  Centre Commercial Sainte Catherine         NaN   
4795    -79.391052                              Hermes         NaN   
...            ...                                 ...         ...   
181148  -79.996915           Smithfield Liberty Garage         NaN   
184673  -79.300685

We have 0 business without a state, 1 business without a city, 659 businesses without a postal_code, 7682 businesses without an address among which only 545 has no postal_code, so we have 114 businesses with an address but no postal_code

In [604]:
eval(business_raw["attributes"][2])

{'GoodForMeal': "{'dessert': False, 'latenight': False, 'lunch': True, 'dinner': True, 'brunch': False, 'breakfast': False}",
 'BusinessParking': "{'garage': False, 'street': False, 'validated': False, 'lot': True, 'valet': False}",
 'NoiseLevel': "u'loud'",
 'Ambience': "{'romantic': False, 'intimate': False, 'classy': False, 'hipster': False, 'divey': False, 'touristy': False, 'trendy': False, 'upscale': False, 'casual': True}"}


# Parse businesses ids

In [605]:
business_ids = business_raw["business_id"].reset_index().set_index("business_id")["index"].to_dict()
del business_raw["business_id"]
business_table = business_raw.reset_index().rename(columns={'index':'id'})
business_table

id                                   address  \
0            1               2818 E Camino Acequia Drive   
1            2                      30 Eglinton Avenue W   
2            3              1210 8th Street SW, Unit 220   
3            4               6955 N Durango Dr, Ste 1116   
4            5                         4 E University Dr   
...        ...                                       ...   
192604  192605                521 St. Catherine Street W   
192605  192606                       244 Victoria Street   
192606  192607                       10210 South 50th Pl   
192607  192608  6670 S Tenaya Way, Ste 120, The Body Spa   
192608  192609                 14050 N 83rd Ave, Ste 290   

                                               attributes  \
0                                                      {}   
1       {'GoodForMeal': "{'dessert': False, 'latenight...   
2       {'BusinessParking': "{'garage': False, 'street...   
3       {'BusinessParking': "{'garage': False, 'street...   
4                                                      {}   
...                                                   ...   
192604  {'Ambience': "{'romantic': False, 'intimate': ...   
192605                                                 {}   
192606                                                NaN   
192607  {'BusinessParking': "{'garage': False, 'street...   
192608                                                NaN   

                                               categories         city  \
0                                       Golf, Active Life      Phoenix   
1       Specialty Food, Restaurants, Dim Sum, Imported...  Mississauga   
2       Local Services, Professional Services, Compute...      Calgary   
3                              Beauty & Spas, Hair Salons    Las Vegas   
4                     American (Traditional), Restaurants        Tempe   
...                                                   ...          ...   
192604            Restaurants, Venezuelan, Latin American     Montréal   
192605              Arts & Entertainment, Performing Arts      Toronto   
192606           Education, Middle Schools & High Schools      Phoenix   
192607  Beauty & Spas, Eyelash Service, Waxing, Esthet...    Las Vegas   
192608  Professional Services, Wills, Trusts, & Probat...       Peoria   

                                                    hours  is_open   latitude  \
0                                                     NaN        0  33.522143   
1       {'Monday': '9:0-0:0', 'Tuesday': '9:0-0:0', 'W...        1  43.605499   
2       {'Monday': '9:0-17:0', 'Tuesday': '9:0-17:0', ...        1  51.041771   
3       {'Monday': '8:30-20:0', 'Tuesday': '8:30-20:0'...        1  36.287312   
4                                                     NaN        0  33.422192   
...                                                   ...      ...        ...   
192604                                                NaN        0  45.504597   
192605                                                NaN        1  43.655290   
192606  {'Monday': '7:30-17:30', 'Tuesday': '7:30-17:3...        1  33.354284   
192607  {'Monday': '0:0-0:0', 'Tuesday': '9:0-19:0', '...        1  36.066952   
192608  {'Monday': '8:0-18:0', 'Tuesday': '8:0-18:0', ...        1  33.612226   

         longitude                         name postal_code  review_count  \
0      -112.018481   Arizona Biltmore Golf Club       85016             5   
1       -79.652289   Emerald Chinese Restaurant     L5R 3E7           128   
2      -114.081109  Nucleus Information Service     T2R 1L3             5   
3      -115.289540                 Amazing Cuts       89149           116   
4      -111.939615                 Ruby Tuesday       85281             9   
...            ...                          ...         ...           ...   
192604  -73.568456                       Arepas     H3B 1A6            15   
192605  -79.379541           Ed Mirvish Theatre     M5B 1V8            58   
192

## Fix business postal_code and city

In [606]:
business_table[business_table.postal_code.isna()]

id              address  \
476        477                  NaN   
479        480       Gerrard Square   
492        493                  NaN   
559        560           Market 707   
817        818  Plaza Hotel Main St   
...        ...                  ...   
190136  190137                  NaN   
190889  190890                  NaN   
191417  191418                  NaN   
191920  191921                  NaN   
192075  192076                  NaN   

                                               attributes  \
476                                                   NaN   
479                                                    {}   
492                                                   NaN   
559     {'Ambience': "{'touristy': False, 'hipster': F...   
817                                                    {}   
...                                                   ...   
190136                                                 {}   
190889                                                NaN   
191417                                                 {}   
191920                                                NaN   
192075                                                 {}   

                                               categories       city  \
476                            Home Services, Landscaping    Phoenix   
479     General Dentistry, Dentists, Health & Medical,...    Toronto   
492                               Plumbing, Home Services    Toronto   
559                                Caribbean, Restaurants    Toronto   
817     Arts & Entertainment, Nightlife, Performing Ar...  Las Vegas   
...                                                   ...        ...   
190136                                  Food Trucks, Food   Montréal   
190889                 Ticket Sales, Arts & Entertainment  Las Vegas   
191417   Venues & Event Spaces, Event Planning & Services  Las Vegas   
191920                             Tours, Hotels & Travel    Toronto   
192075  Glass & Mirrors, Windows Installation, Shutter...  Las Vegas   

                                                    hours  is_open   latitude  \
476                                                   NaN        1  33.448377   
479                                                   NaN        1  43.670630   
492                                                   NaN        0  43.653226   
559                                                   NaN        1  43.652196   
817                                                   NaN        1  36.180670   
...                                                   ...      ...        ...   
190136                                                NaN        1  45.508670   
190889                                                NaN        1  36.169941   
191417  {'Tuesday': '10:0-18:0', 'Wednesday': '10:0-18...        0  36.175000   
191920  {'Monday': '6:0-23:0', 'Tuesday': '6:0-23:0', ...        1  43.653226   
192075  {'Monday': '9:0-17:0', 'Tuesday': '9:0-17:0', ...        1  36.169941   

         longitude                              name postal_code  \
476    -112.074037                Robert Landscaping         NaN   
479     -79.329283                  Riverdale Dental         NaN   
492     -79.383184                      Rapid Rooter         NaN   
559     -79.405563  T&T Roti Trinidad & Toronto Roti         NaN   
817    -115.140494                       Avant Garde         NaN   
...            ...                               ...         ...   
190136  -73.553993                        Chaud Dogs         NaN   
190889 -115.139830             NoHidden Fees Tickets         NaN   
191417 -115.136389              The Griffin Mansions         NaN   
191920  -79.383184            Niagara Falls Day Tour         NaN   
192075 -115.139830                    Shutters Cheap         NaN   

        review_count  stars state  
476                4    1.0    AZ  
479                5    2.5    ON  
492                3    1.0    ON  
559        

### Locators used

In [625]:
locator = Nominatim(user_agent="myGeocoder",timeout=3)
location = locator.reverse("36.169710,-115.123695")
location.raw

{'place_id': 177667178,
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright',
 'osm_type': 'way',
 'osm_id': 395751948,
 'lat': '36.16966995091101',
 'lon': '-115.12369453339117',
 'display_name': 'Cultural Corridor, Five Points, Las Vegas, Clark County, Nevada, NV 89101, United States of America',
 'address': {'neighbourhood': 'Cultural Corridor',
  'suburb': 'Five Points',
  'city': 'Las Vegas',
  'county': 'Clark County',
  'state': 'Nevada',
  'postcode': 'NV 89101',
  'country': 'United States of America',
  'country_code': 'us'},
 'boundingbox': ['36.1696534', '36.1697054', '-115.1251151', '-115.120655']}

In [608]:
mapQuestLocator=geopy.geocoders.OpenMapQuest("StxlGpGLb5EapoCXFQBf6GroFDOZTBJj",timeout=3)
location=coder.reverse("36.169710,-115.123695")
location.raw

{'place_id': '158781428',
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright',
 'osm_type': 'way',
 'osm_id': '395751948',
 'lat': '36.1697054',
 'lon': '-115.120655',
 'display_name': 'Cultural Corridor, Five Points, Las Vegas, Clark County, Nevada, 89102, United States of America',
 'address': {'neighbourhood': 'Cultural Corridor',
  'suburb': 'Five Points',
  'city': 'Las Vegas',
  'county': 'Clark County',
  'state': 'Nevada',
  'postcode': '89102',
  'country': 'United States of America',
  'country_code': 'us'},
 'boundingbox': ['36.1696534', '36.1697328', '-115.1251151', '-115.1160606']}

In [609]:
openCageLocator=geopy.geocoders.OpenCage("71ef46faea4b4e198a0891d79714905b",timeout=3)
location=coder.reverse("36.169710,-115.123695")
location.raw  

{'place_id': '158781428',
 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright',
 'osm_type': 'way',
 'osm_id': '395751948',
 'lat': '36.1697054',
 'lon': '-115.120655',
 'display_name': 'Cultural Corridor, Five Points, Las Vegas, Clark County, Nevada, 89102, United States of America',
 'address': {'neighbourhood': 'Cultural Corridor',
  'suburb': 'Five Points',
  'city': 'Las Vegas',
  'county': 'Clark County',
  'state': 'Nevada',
  'postcode': '89102',
  'country': 'United States of America',
  'country_code': 'us'},
 'boundingbox': ['36.1696534', '36.1697328', '-115.1251151', '-115.1160606']}

In [ ]:
googleLocator=geopy.geocoders.GoogleV3(api_key=YOUR_API_KEY)

In [694]:
business_table[business_table.address.isna()]

Empty DataFrame
Columns: [id, address, attributes, categories, city, hours, is_open, latitude, longitude, name, postal_code, review_count, stars, state]
Index: []

In [631]:
business_table[business_table.city.isna()]

Empty DataFrame
Columns: [id, address, attributes, categories, city, hours, is_open, latitude, longitude, name, postal_code, review_count, stars, state]
Index: []

In [695]:
business_table[business_table.postal_code.isna()]

Empty DataFrame
Columns: [id, address, attributes, categories, city, hours, is_open, latitude, longitude, name, postal_code, review_count, stars, state]
Index: []

In [632]:
def get_postal_code(data):
    coordinates = f'{data.latitude},{data.longitude}'
    location = mapQuestLocator.reverse(coordinates)
    if location.raw.get("address").get("postcode") == None:
        location = locator.reverse(coordinates)
    return location.raw.get("address").get("postcode")

In [651]:
def get_postal_code_v2(data):
    coordinates = f'{data.latitude},{data.longitude}'
    location = googleLocator.reverse(coordinates)
    return list(filter(lambda l:l.get("types")[0]=="postal_code",location[0].raw.get("address_components")))[0].get("long_name")

In [635]:
def get_city_name(data):
    coordinates = f'{data.latitude},{data.longitude}'
    location = locator.reverse(coordinates)
    if location.raw.get("address").get("city") == None:
        location=mapQuestLocator.reverse(coordinates)
    return location.raw.get("address").get("city",np.NaN)

In [426]:
def get_address(data):
    coordinates = f'{data.latitude},{data.longitude}'
    location = locator.reverse(coordinates)
    if location.raw.get("address").get("road") == None or location.raw.get("neighbourhood") == None:
        location=mapQuestLocator.reverse(coordinates)
    return location.raw.get("address").get("road", location.raw.get("address").get("neighbourhood", np.NaN))

In [692]:
def format(locs):
    res=np.NaN
    route=list(filter(lambda x: x.get("types")[0]=="route", locs.get("address_components")))
    nb=list(filter(lambda x: x.get("types")[0]=="street_number", locs.get("address_components")))
    neighborhood=list(filter(lambda x: x.get("types")[0]=="neighborhood", locs.get("address_components")))
    if(len(route)>0):
        res=route[0].get("long_name")
    if(not np.isnan(res) and len(nb)>0):
        res=res+" "+nb[0].get("long_name")
    if(np.isnan(res) and len(neighborhood)>0):
        res=neighborhood[0].get("long_name")
    return res
def get_address_v2(data):
    coordinates = f'{data.latitude},{data.longitude}'
    locations = googleLocator.reverse(coordinates)
    res=format(location[0].raw)
    if np.isnan(res) and len(locations)>0:
        res=format(location[1].raw)
    return res

In [633]:
business_table.loc[business_table.postal_code.isna(),'postal_code'] = business_table[business_table.postal_code.isna()].progress_apply(get_postal_code, axis=1)

100%|██████████| 3/3 [00:01<00:00,  1.55it/s]


In [630]:
business_table.loc[business_table.city.isna(),'city'] = business_table[business_table.city.isna()].progress_apply(get_city_name, axis=1)

100%|██████████| 1/1 [00:00<00:00,  6.07it/s]


In [638]:
business_table.loc[business_table.address.isna(),'address'] = business_table[business_table.address.isna()].progress_apply(get_address, axis=1)

100%|██████████| 7682/7682 [1:20:59<00:00,  1.58it/s]


In [684]:
business_table.loc[business_table.postal_code.isna(),'postal_code'] = business_table[business_table.postal_code.isna()].progress_apply(get_postal_code_v2, axis=1)

  0%|          | 0/1 [00:00<?, ?it/s]/home/bastien/.local/lib/python3.6/site-packages/ipykernel_launcher.py:3: DeprecationWarning: GoogleV3.reverse: default value for `exactly_one` argument will become True in geopy 2.0. Specify `exactly_one=False` as the argument explicitly to get rid of this warning.
  This is separate from the ipykernel package so we can avoid doing imports until
100%|██████████| 1/1 [00:00<00:00,  7.03it/s]


In [693]:
business_table.loc[business_table.address.isna(),'address'] = business_table[business_table.address.isna()].progress_apply(get_address_v2, axis=1)

  0%|          | 0/175 [00:00<?, ?it/s]/home/bastien/.local/lib/python3.6/site-packages/ipykernel_launcher.py:15: DeprecationWarning: GoogleV3.reverse: default value for `exactly_one` argument will become True in geopy 2.0. Specify `exactly_one=False` as the argument explicitly to get rid of this warning.
  from ipykernel import kernelapp as app
100%|██████████| 175/175 [00:18<00:00,  9.42it/s]


# Parse Locations

In [697]:
locations = business_table[["id", "address", "city", "latitude", "longitude", "postal_code", "state"]]
del business_table["address"]
del business_table["city"]
del business_table["latitude"]
del business_table["longitude"]
del business_table["postal_code"]
del business_table["state"]
locations = locations.rename(columns={'id':'business_id'})
locations

business_id                                   address         city  \
0                 1               2818 E Camino Acequia Drive      Phoenix   
1                 2                      30 Eglinton Avenue W  Mississauga   
2                 3              1210 8th Street SW, Unit 220      Calgary   
3                 4               6955 N Durango Dr, Ste 1116    Las Vegas   
4                 5                         4 E University Dr        Tempe   
...             ...                                       ...          ...   
192604       192605                521 St. Catherine Street W     Montréal   
192605       192606                       244 Victoria Street      Toronto   
192606       192607                       10210 South 50th Pl      Phoenix   
192607       192608  6670 S Tenaya Way, Ste 120, The Body Spa    Las Vegas   
192608       192609                 14050 N 83rd Ave, Ste 290       Peoria   

         latitude   longitude postal_code state  
0       33.522143 -112.018481       85016    AZ  
1       43.605499  -79.652289     L5R 3E7    ON  
2       51.041771 -114.081109     T2R 1L3    AB  
3       36.287312 -115.289540       89149    NV  
4       33.422192 -111.939615       85281    AZ  
...           ...         ...         ...   ...  
192604  45.504597  -73.568456     H3B 1A6    QC  
192605  43.655290  -79.379541     M5B 1V8    ON  
192606  33.354284 -111.975538       85044    AZ  
192607  36.066952 -115.251008       89113    NV  
192608  33.612226 -112.238145       85381    AZ  

[192609 rows x 7 columns]

## States

In [698]:
state_table = locations["state"].drop_duplicates()
state_table = state_table.sort_values().reset_index()
state_table.index += 1
del state_table["index"]
state_table = state_table.reset_index().rename(columns={"index":"id", "state": "name"})
state_table

id name
0    1   AB
1    2   AK
2    3   AL
3    4   AR
4    5   AZ
5    6  BAS
6    7   BC
7    8   CA
8    9  CON
9   10   CT
10  11  DOW
11  12  DUR
12  13   FL
13  14   GA
14  15   IL
15  16   NC
16  17   NE
17  18   NJ
18  19   NM
19  20   NV
20  21   NY
21  22   OH
22  23   ON
23  24   PA
24  25   QC
25  26   SC
26  27   TN
27  28   TX
28  29   UT
29  30   VA
30  31   VT
31  32   WA
32  33   WI
33  34  XGL
34  35  XGM
35  36  XWY

In [699]:
state_dict = state_table.set_index("name")["id"].to_dict()
locations["state_id"] = locations["state"].progress_apply(lambda s: state_dict[s])
del locations["state"]
locations

100%|██████████| 192609/192609 [00:00<00:00, 443180.64it/s]


business_id                                   address         city  \
0                 1               2818 E Camino Acequia Drive      Phoenix   
1                 2                      30 Eglinton Avenue W  Mississauga   
2                 3              1210 8th Street SW, Unit 220      Calgary   
3                 4               6955 N Durango Dr, Ste 1116    Las Vegas   
4                 5                         4 E University Dr        Tempe   
...             ...                                       ...          ...   
192604       192605                521 St. Catherine Street W     Montréal   
192605       192606                       244 Victoria Street      Toronto   
192606       192607                       10210 South 50th Pl      Phoenix   
192607       192608  6670 S Tenaya Way, Ste 120, The Body Spa    Las Vegas   
192608       192609                 14050 N 83rd Ave, Ste 290       Peoria   

         latitude   longitude postal_code  state_id  
0       33.522143 -112.018481       85016         5  
1       43.605499  -79.652289     L5R 3E7        23  
2       51.041771 -114.081109     T2R 1L3         1  
3       36.287312 -115.289540       89149        20  
4       33.422192 -111.939615       85281         5  
...           ...         ...         ...       ...  
192604  45.504597  -73.568456     H3B 1A6        25  
192605  43.655290  -79.379541     M5B 1V8        23  
192606  33.354284 -111.975538       85044         5  
192607  36.066952 -115.251008       89113        20  
192608  33.612226 -112.238145       85381         5  

[192609 rows x 7 columns]

## Cities

In [700]:
city_table = locations[["city", "state_id"]].drop_duplicates()
city_table = city_table.reset_index()
del city_table["index"]
city_table.index += 1
city_table = city_table.reset_index()
city_table = city_table.rename(columns={"index":"id", "city":"name"})
city_table

id         name  state_id
0        1      Phoenix         5
1        2  Mississauga        23
2        3      Calgary         1
3        4    Las Vegas        20
4        5        Tempe         5
...    ...          ...       ...
1252  1253      Pahrump        20
1253  1254    ETOBICOKE        23
1254  1255  Rockey View         1
1255  1256   Pittsburch        24
1256  1257  Richmonhill        23

[1257 rows x 3 columns]

In [701]:
city_dict = city_table.set_index(["name", "state_id"])["id"].to_dict()

In [702]:
locations["city_id"] = locations[["city", "state_id"]].progress_apply(lambda row: city_dict[(row[0], row[1])], axis=1, raw=True)
del locations["state_id"]
del locations["city"]
locations

100%|██████████| 192609/192609 [00:06<00:00, 28890.55it/s]


business_id                                   address   latitude  \
0                 1               2818 E Camino Acequia Drive  33.522143   
1                 2                      30 Eglinton Avenue W  43.605499   
2                 3              1210 8th Street SW, Unit 220  51.041771   
3                 4               6955 N Durango Dr, Ste 1116  36.287312   
4                 5                         4 E University Dr  33.422192   
...             ...                                       ...        ...   
192604       192605                521 St. Catherine Street W  45.504597   
192605       192606                       244 Victoria Street  43.655290   
192606       192607                       10210 South 50th Pl  33.354284   
192607       192608  6670 S Tenaya Way, Ste 120, The Body Spa  36.066952   
192608       192609                 14050 N 83rd Ave, Ste 290  33.612226   

         longitude postal_code  city_id  
0      -112.018481       85016        1  
1       -79.652289     L5R 3E7        2  
2      -114.081109     T2R 1L3        3  
3      -115.289540       89149        4  
4      -111.939615       85281        5  
...            ...         ...      ...  
192604  -73.568456     H3B 1A6        7  
192605  -79.379541     M5B 1V8       23  
192606 -111.975538       85044        1  
192607 -115.251008       89113        4  
192608 -112.238145       85381       35  

[192609 rows x 6 columns]

## Postal codes

In [703]:
postal_codes_table = locations[["city_id", "postal_code"]].drop_duplicates()
postal_codes_table = postal_codes_table.reset_index()
del postal_codes_table["index"]
postal_codes_table.index += 1
postal_codes_table = postal_codes_table.reset_index().rename(columns={"index":"id"})
postal_codes_table

id  city_id postal_code
0          1        1       85016
1          2        2     L5R 3E7
2          3        3     T2R 1L3
3          4        4       89149
4          5        5       85281
...      ...      ...         ...
20429  20430      164       44128
20430  20431       16     M1E 4P7
20431  20432      536     J7Y 3L8
20432  20433        7     H3H 1Y2
20433  20434      695     J3V 6H1

[20434 rows x 3 columns]

In [704]:
postal_code_dict = postal_codes_table.set_index(["postal_code", "city_id"])["id"].to_dict()

In [705]:
locations["postal_code_id"] = locations[["postal_code", "city_id"]].progress_apply(lambda row: postal_code_dict[(row[0], row[1])], axis=1, raw=True)
del locations["city_id"]
del locations["postal_code"]
locations

100%|██████████| 192609/192609 [00:06<00:00, 28748.07it/s]


business_id                                   address   latitude  \
0                 1               2818 E Camino Acequia Drive  33.522143   
1                 2                      30 Eglinton Avenue W  43.605499   
2                 3              1210 8th Street SW, Unit 220  51.041771   
3                 4               6955 N Durango Dr, Ste 1116  36.287312   
4                 5                         4 E University Dr  33.422192   
...             ...                                       ...        ...   
192604       192605                521 St. Catherine Street W  45.504597   
192605       192606                       244 Victoria Street  43.655290   
192606       192607                       10210 South 50th Pl  33.354284   
192607       192608  6670 S Tenaya Way, Ste 120, The Body Spa  36.066952   
192608       192609                 14050 N 83rd Ave, Ste 290  33.612226   

         longitude  postal_code_id  
0      -112.018481               1  
1       -79.652289               2  
2      -114.081109               3  
3      -115.289540               4  
4      -111.939615               5  
...            ...             ...  
192604  -73.568456            5509  
192605  -79.379541           13253  
192606 -111.975538             205  
192607 -115.251008             260  
192608 -112.238145             123  

[192609 rows x 5 columns]

# Parse categories

In [752]:
business_categories_temp = business_table[["id", "categories"]][business_table["categories"].progress_map(lambda x: type(x) == str)]
business_categories_temp["categories"] = business_categories_temp["categories"].progress_map(lambda cats: list(map(str.strip, cats.split(","))))

business_categories_temp

KeyError: "['categories'] not in index"

In [753]:
business_categories = (business_categories_temp['categories']
    .progress_apply(lambda x: pd.Series(x))
    .stack()
    .reset_index(level=1, drop=True)
    .to_frame('categories')
    .join(business_categories_temp[['id']], how='left'))
business_categories

100%|██████████| 192127/192127 [01:15<00:00, 2553.60it/s]


categories      id
0                      Golf       1
0               Active Life       1
1            Specialty Food       2
1               Restaurants       2
1                   Dim Sum       2
...                     ...     ...
192608               Trusts  192609
192608           & Probates  192609
192608              Lawyers  192609
192608       Legal Services  192609
192608  Estate Planning Law  192609

[788359 rows x 2 columns]

In [754]:
categories = business_categories["categories"]
categories = categories.drop_duplicates().reset_index()
del categories["index"]
categories.index += 1
categories = categories.reset_index().rename(columns={"index":"id", "categories": "name"})
categories

id               name
0        1               Golf
1        2        Active Life
2        3     Specialty Food
3        4        Restaurants
4        5            Dim Sum
...    ...                ...
1295  1296         Island Pub
1296  1297        Homeopathic
1297  1298  Medical Foot Care
1298  1299        Geneticists
1299  1300        Calligraphy

[1300 rows x 2 columns]

In [760]:
categories_dict = categories.reset_index().set_index("name")["index"].to_dict()
business_categories["categorie_id"] = business_categories["categories"].progress_map(lambda cat: categories_dict[cat])
del business_categories["categories"]
del business_table["categories"]
business_categories = business_categories.rename(columns={"id":"business_id"}).drop_duplicates()
business_categories

KeyError: 'categories'

# Parse attributes

In [710]:
attributes = business_raw[business_raw["attributes"].map(lambda x: type(x)==str)]["attributes"].progress_apply(eval)
del business_table["attributes"]
attributes

100%|██████████| 163773/163773 [00:01<00:00, 97203.57it/s] 


1                                                        {}
2         {'GoodForMeal': '{'dessert': False, 'latenight...
3         {'BusinessParking': '{'garage': False, 'street...
4         {'BusinessParking': '{'garage': False, 'street...
5                                                        {}
                                ...                        
192602                                                   {}
192604                                                   {}
192605    {'Ambience': '{'romantic': False, 'intimate': ...
192606                                                   {}
192608    {'BusinessParking': '{'garage': False, 'street...
Name: attributes, Length: 163773, dtype: object

## Noise level

In [761]:
noise_level_temp = attributes[attributes.map(lambda x: "NoiseLevel" in x)].map(lambda x: eval(x["NoiseLevel"]))
noise_level = noise_level_temp.drop_duplicates().reset_index()
del noise_level["index"]
noise_level.index += 1
noise_level = noise_level.reset_index().rename(columns={"index":"id", "attributes":"level"})
noise_level

id      level
0   1       loud
1   2    average
2   3      quiet
3   4  very_loud
4   5       None

In [762]:
noise_level_dict = noise_level.set_index("level")["id"].to_dict()
business_table["noise_level_id"] = noise_level_temp.map(lambda x: noise_level_dict[x])
#uses that NaN != NaN
business_table["noise_level_id"] = business_table["noise_level_id"].apply(lambda x: int(x) if x == x else "")
#ugly but working for now
business_table["noise_level_id"] = business_table["noise_level_id"].apply(lambda x: x if x != 5 else "")
noise_level = noise_level.dropna()
business_table

id  is_open                         name  review_count  stars  \
0            1        0   Arizona Biltmore Golf Club             5    3.0   
1            2        1   Emerald Chinese Restaurant           128    2.5   
2            3        1  Nucleus Information Service             5    2.0   
3            4        1                 Amazing Cuts           116    4.5   
4            5        0                 Ruby Tuesday             9    2.5   
...        ...      ...                          ...           ...    ...   
192604  192605        0                       Arepas            15    3.5   
192605  192606        1           Ed Mirvish Theatre            58    4.0   
192606  192607        1              BASIS Ahwatukee             3    2.5   
192607  192608        1              Brows Redefined            15    5.0   
192608  192609        1   Affordable Estate Planning             3    5.0   

       noise_level_id  
0                      
1                      
2                   1  
3                      
4                      
...               ...  
192604                 
192605              2  
192606                 
192607                 
192608                 

[192609 rows x 6 columns]

## Music

In [763]:
music_temp = attributes[attributes.map(lambda x: "Music" in x)].map(lambda x: eval(x["Music"]))
music_temp = music_temp[music_temp.map(lambda x: x != None)]
music_temp = music_temp.progress_map(lambda m: [key for key in m.keys() if m[key]])
music_temp = music_temp[music_temp.map(lambda l: len(l) > 0)].reset_index().rename(columns={"index":"business_id", "attributes":"name"})
music_temp

100%|██████████| 5181/5181 [00:00<00:00, 248666.18it/s]


business_id                       name
0              94                  [jukebox]
1             140                       [dj]
2             164  [background_music, video]
3             207         [background_music]
4             255         [background_music]
...           ...                        ...
2376       192154         [background_music]
2377       192182                  [karaoke]
2378       192394         [background_music]
2379       192408                  [jukebox]
2380       192595                     [live]

[2381 rows x 2 columns]

In [764]:
music_business_relation_table = (music_temp['name']
    .progress_apply(lambda x: pd.Series(x))
    .stack()
    .reset_index(level=1, drop=True)
    .to_frame('name')
    .join(music_temp[['business_id']], how='left'))
music_business_relation_table

100%|██████████| 2381/2381 [00:00<00:00, 2742.45it/s]


name  business_id
0              jukebox           94
1                   dj          140
2     background_music          164
2                video          164
3     background_music          207
...                ...          ...
2376  background_music       192154
2377           karaoke       192182
2378  background_music       192394
2379           jukebox       192408
2380              live       192595

[2759 rows x 2 columns]

In [765]:
music_table = music_business_relation_table["name"].drop_duplicates().reset_index()
del music_table["index"]
music_table.index += 1
music_table = music_table.reset_index().rename(columns={"index":"id"})
music_table

id              name
0   1           jukebox
1   2                dj
2   3  background_music
3   4             video
4   5              live
5   6           karaoke

In [766]:
music_dict = music_table.set_index("name")["id"].to_dict()
music_business_relation_table["music_id"] = music_business_relation_table["name"].map(lambda x: music_dict[x])
del music_business_relation_table["name"]
music_business_relation_table

business_id  music_id
0              94         1
1             140         2
2             164         3
2             164         4
3             207         3
...           ...       ...
2376       192154         3
2377       192182         6
2378       192394         3
2379       192408         1
2380       192595         5

[2759 rows x 2 columns]

## Business parking 

In [767]:
parking_temp = attributes[attributes.map(lambda x: "BusinessParking" in x)].map(lambda x: eval(x["BusinessParking"]))
parking_temp = parking_temp[parking_temp.map(lambda x: x != None)]
parking_temp = parking_temp.progress_map(lambda m: [key for key in m.keys() if m[key]])
parking_temp = parking_temp[parking_temp.map(lambda l: len(l) > 0)].reset_index().rename(columns={"index":"business_id", "attributes":"name"})
parking_temp

100%|██████████| 101478/101478 [00:00<00:00, 237818.03it/s]


business_id      name
0                2     [lot]
1                3  [street]
2                4     [lot]
3               12     [lot]
4               22     [lot]
...            ...       ...
61241       192594     [lot]
61242       192599  [garage]
61243       192600     [lot]
61244       192601  [street]
61245       192608     [lot]

[61246 rows x 2 columns]

In [768]:
parking_business_relation_table = (parking_temp['name']
    .progress_apply(lambda x: pd.Series(x))
    .stack()
    .reset_index(level=1, drop=True)
    .to_frame('name')
    .join(parking_temp[['business_id']], how='left'))
parking_business_relation_table

100%|██████████| 61246/61246 [00:22<00:00, 2664.30it/s]


name  business_id
0         lot            2
1      street            3
2         lot            4
3         lot           12
4         lot           22
...       ...          ...
61241     lot       192594
61242  garage       192599
61243     lot       192600
61244  street       192601
61245     lot       192608

[69363 rows x 2 columns]

In [769]:
business_parking_table = parking_business_relation_table["name"].drop_duplicates().reset_index()
business_parking_table.index += 1
del business_parking_table["index"]
business_parking_table = business_parking_table.reset_index().rename(columns={"index":"id"})
business_parking_table

id       name
0   1        lot
1   2     street
2   3     garage
3   4      valet
4   5  validated

In [770]:
parking_dict = business_parking_table.set_index("name")["id"].to_dict()
parking_business_relation_table["parking_id"] = parking_business_relation_table["name"].map(lambda x: parking_dict[x])
del parking_business_relation_table["name"]
parking_business_relation_table

business_id  parking_id
0                2           1
1                3           2
2                4           1
3               12           1
4               22           1
...            ...         ...
61241       192594           1
61242       192599           3
61243       192600           1
61244       192601           2
61245       192608           1

[69363 rows x 2 columns]

## Ambience

In [771]:
ambience_temp = attributes[attributes.map(lambda x: "Ambience" in x)].map(lambda x: eval(x["Ambience"]))
ambience_temp = ambience_temp[ambience_temp.map(lambda x: x != None)]
ambience_temp = ambience_temp.progress_map(lambda m: [key for key in m.keys() if m[key]])
ambience_temp = ambience_temp[ambience_temp.map(lambda l: len(l) > 0)].reset_index().rename(columns={"index":"business_id", "attributes":"name"})
ambience_temp

100%|██████████| 47547/47547 [00:00<00:00, 227923.64it/s]


business_id                name
0                2            [casual]
1               19            [casual]
2               32            [casual]
3               50            [casual]
4               59            [casual]
...            ...                 ...
26701       192567  [touristy, casual]
26702       192571            [casual]
26703       192580            [casual]
26704       192593            [casual]
26705       192596            [casual]

[26706 rows x 2 columns]

In [772]:
ambience_business_relation_table = (ambience_temp['name']
    .progress_apply(lambda x: pd.Series(x))
    .stack()
    .reset_index(level=1, drop=True)
    .to_frame('name')
    .join(ambience_temp[['business_id']], how='left'))
ambience_business_relation_table

100%|██████████| 26706/26706 [00:10<00:00, 2652.50it/s]


name  business_id
0      casual            2
1      casual           19
2      casual           32
3      casual           50
4      casual           59
...       ...          ...
26701  casual       192567
26702  casual       192571
26703  casual       192580
26704  casual       192593
26705  casual       192596

[30739 rows x 2 columns]

In [773]:
ambience_table = ambience_business_relation_table["name"].drop_duplicates().reset_index()
ambience_table.index += 1
del ambience_table["index"]
ambience_table = ambience_table.reset_index().rename(columns={"index":"id"})
ambience_table

id      name
0   1    casual
1   2   hipster
2   3    trendy
3   4  romantic
4   5  intimate
5   6    classy
6   7     divey
7   8  touristy
8   9   upscale

In [774]:
ambience_dict = ambience_table.set_index("name")["id"].to_dict()
ambience_business_relation_table["ambience_id"] = ambience_business_relation_table["name"].map(lambda x: ambience_dict[x])
del ambience_business_relation_table["name"]
ambience_business_relation_table

business_id  ambience_id
0                2            1
1               19            1
2               32            1
3               50            1
4               59            1
...            ...          ...
26701       192567            1
26702       192571            1
26703       192580            1
26704       192593            1
26705       192596            1

[30739 rows x 2 columns]

## Good for meal

In [775]:
good_for_meal_temp = attributes[attributes.map(lambda x: "GoodForMeal" in x)].map(lambda x: eval(x["GoodForMeal"]))
good_for_meal_temp = good_for_meal_temp[good_for_meal_temp.map(lambda x: x != None)]
good_for_meal_temp = good_for_meal_temp.progress_map(lambda m: [key for key in m.keys() if m[key]])
good_for_meal_temp = good_for_meal_temp[good_for_meal_temp.map(lambda l: len(l) > 0)].reset_index().rename(columns={"index":"business_id", "attributes":"name"})
good_for_meal_temp

100%|██████████| 29715/29715 [00:00<00:00, 259985.11it/s]


business_id                        name
0                2             [lunch, dinner]
1               19             [lunch, dinner]
2               32             [lunch, dinner]
3               59    [dessert, lunch, dinner]
4               64  [lunch, brunch, breakfast]
...            ...                         ...
21251       192548                    [dinner]
21252       192567             [lunch, brunch]
21253       192571                     [lunch]
21254       192580             [lunch, dinner]
21255       192596                 [breakfast]

[21256 rows x 2 columns]

In [776]:
good_for_meal_business_relation_table = (good_for_meal_temp['name']
    .progress_apply(lambda x: pd.Series(x))
    .stack()
    .reset_index(level=1, drop=True)
    .to_frame('name')
    .join(good_for_meal_temp[['business_id']], how='left'))
good_for_meal_business_relation_table

100%|██████████| 21256/21256 [00:08<00:00, 2428.75it/s]


name  business_id
0          lunch            2
0         dinner            2
1          lunch           19
1         dinner           19
2          lunch           32
...          ...          ...
21252     brunch       192567
21253      lunch       192571
21254      lunch       192580
21254     dinner       192580
21255  breakfast       192596

[41047 rows x 2 columns]

In [777]:
good_for_meal_table = good_for_meal_business_relation_table["name"].drop_duplicates().reset_index()
good_for_meal_table.index += 1
del good_for_meal_table["index"]
good_for_meal_table = good_for_meal_table.reset_index().rename(columns={"index":"id"})
good_for_meal_table

id       name
0   1      lunch
1   2     dinner
2   3    dessert
3   4     brunch
4   5  breakfast
5   6  latenight

In [778]:
good_for_meal_dict = good_for_meal_table.set_index("name")["id"].to_dict()
good_for_meal_business_relation_table["ambience_id"] = good_for_meal_business_relation_table["name"].map(lambda x: good_for_meal_dict[x])
del good_for_meal_business_relation_table["name"]
good_for_meal_business_relation_table

business_id  ambience_id
0                2            1
0                2            2
1               19            1
1               19            2
2               32            1
...            ...          ...
21252       192567            4
21253       192571            1
21254       192580            1
21254       192580            2
21255       192596            5

[41047 rows x 2 columns]

## dietary restrictions

In [779]:
dietary_restrictions_temp = attributes[attributes.map(lambda x: "DietaryRestrictions" in x)].map(lambda x: eval(x["DietaryRestrictions"]))
dietary_restrictions_temp = dietary_restrictions_temp[dietary_restrictions_temp.map(lambda x: x != None)]
dietary_restrictions_temp = dietary_restrictions_temp.progress_map(lambda m: [key for key in m.keys() if m[key]])
dietary_restrictions_temp = dietary_restrictions_temp[dietary_restrictions_temp.map(lambda l: len(l) > 0)].reset_index().rename(columns={"index":"business_id", "attributes":"name"})
dietary_restrictions_temp

100%|██████████| 52/52 [00:00<00:00, 22905.25it/s]


business_id                                               name
0           963                                      [gluten-free]
1          2091                                            [halal]
2          3777                                      [gluten-free]
3         10048                                            [vegan]
4         12861                                       [vegetarian]
5         13187                    [dairy-free, vegan, vegetarian]
6         14766                               [gluten-free, vegan]
7         16687  [dairy-free, gluten-free, vegan, soy-free, veg...
8         16942                          [gluten-free, vegetarian]
9         22468                                      [gluten-free]
10        25002                                       [vegetarian]
11        25557                                       [vegetarian]
12        26647                                       [vegetarian]
13        45225                                      [gluten-free]
14        45630                                            [vegan]
15        47167                                      [gluten-free]
16        47902                                      [gluten-free]
17        49442                          [gluten-free, vegetarian]
18        49508                                [vegan, vegetarian]
19        51681                               [gluten-free, vegan]
20        75365                   [gluten-free, vegan, vegetarian]
21        77006                                      [gluten-free]
22        89597                                            [vegan]
23        91065                                      [gluten-free]
24        91489                                      [gluten-free]
25        95328                                [vegan, vegetarian]
26        95706                                            [halal]
27        98370                               [gluten-free, vegan]
28        99039                               [gluten-free, vegan]
29       100229                                            [halal]
30       101232                   [gluten-free, vegan, vegetarian]
31       107730                                       [vegetarian]
32       114489                                            [vegan]
33       121881                                            [vegan]
34       123351                                            [vegan]
35       123653                                            [vegan]
36       124378                                      [gluten-free]
37       125085                                            [vegan]
38       125435                                            [vegan]
39       126357                                      [gluten-free]
40       129980                                      [gluten-free]
41       133448                                       [vegetarian]
42       139543                               [gluten-free, vegan]
43       139557                                            [vegan]
44       150309                                [vegan, vegetarian]
45       152386                               [gluten-free, vegan]
46       154652                                            [vegan]
47       158370                                       [vegetarian]
48       158889                                            [vegan]
49       161767                                [vegan, vegetarian]
50       177625                                      [gluten-free]
51       183531                                            [vegan]

In [780]:
dietary_restrictions_business_relation_table = (dietary_restrictions_temp['name']
    .progress_apply(lambda x: pd.Series(x))
    .stack()
    .reset_index(level=1, drop=True)
    .to_frame('name')
    .join(dietary_restrictions_temp[['business_id']], how='left'))
dietary_restrictions_business_relation_table

100%|██████████| 52/52 [00:00<00:00, 2269.03it/s]


name  business_id
0   gluten-free          963
1         halal         2091
2   gluten-free         3777
3         vegan        10048
4    vegetarian        12861
..          ...          ...
48        vegan       158889
49        vegan       161767
49   vegetarian       161767
50  gluten-free       177625
51        vegan       183531

[74 rows x 2 columns]

In [781]:
dietary_restrictions_table = dietary_restrictions_business_relation_table["name"].drop_duplicates().reset_index()
dietary_restrictions_table.index += 1
del dietary_restrictions_table["index"]
dietary_restrictions_table = dietary_restrictions_table.reset_index().rename(columns={"index":"id"})
dietary_restrictions_table

id         name
0   1  gluten-free
1   2        halal
2   3        vegan
3   4   vegetarian
4   5   dairy-free
5   6     soy-free

In [782]:
dietary_restrictions_dict = dietary_restrictions_table.set_index("name")["id"].to_dict()
dietary_restrictions_business_relation_table["dietary_restrictions_id"] = dietary_restrictions_business_relation_table["name"].map(lambda x: dietary_restrictions_dict[x])
del dietary_restrictions_business_relation_table["name"]
dietary_restrictions_business_relation_table

business_id  dietary_restrictions_id
0           963                        1
1          2091                        2
2          3777                        1
3         10048                        3
4         12861                        4
..          ...                      ...
48       158889                        3
49       161767                        3
49       161767                        4
50       177625                        1
51       183531                        3

[74 rows x 2 columns]

# Parse schedule

In [784]:
schedule_temp = business_raw["hours"]
del business_table["hours"]
schedule_temp = schedule_temp[schedule_temp.map(lambda x: type(x) == str)].progress_map(eval)
schedule_temp

KeyError: 'hours'

In [ ]:
days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
day_table = pd.DataFrame(days, columns=["name"])
day_table.index += 1
day_table = day_table.reset_index().rename(columns={"index": "id"})
day_dict = day_table.set_index("name")["id"].to_dict()
day_table

In [ ]:
def schedules_on(day):
    temp = schedule_temp[schedule_temp.map(lambda m: day in m)].progress_map(lambda m: m[day]).reset_index().rename(columns={"index": "business_id"})
    temp["day_id"] = day_dict[day]
    temp["start_at"] = temp["hours"].map(lambda s: s.split("-")[0])
    temp["end_at"] = temp["hours"].map(lambda s: s.split("-")[1])
    del temp["hours"]
    return temp

In [ ]:
schedule_table = pd.concat([schedules_on(day) for day in days])
schedule_table

In [ ]:
business_raw[business_raw["city"] == "New York"]

In [ ]:
locations.loc[119357]

In [ ]:
postal_codes_table[postal_codes_table["id"] == 119358]

In [ ]:
city_table[city_table["id"] == 119358]

In [ ]:
business_raw

In [ ]:
business_table

# Store results

In [785]:
def store_table(table, file_name):
    table.to_csv("generated/{}.csv".format(file_name), index=False)
store_table(state_table, "state")
store_table(city_table, "city")
store_table(postal_codes_table, "postal_code")
store_table(business_table, "business")
store_table(locations, "business_locations")
store_table(categories, "categorie")
store_table(business_categories, "business_categorie")
store_table(noise_level, "noise_level")
store_table(music_table, "music")
store_table(music_business_relation_table, "music_business_relation")
store_table(business_parking_table, "business_parking")
store_table(parking_business_relation_table, "parking_business_relation")
store_table(ambience_table, "ambience")
store_table(ambience_business_relation_table, "ambience_business_relation")
store_table(good_for_meal_table, "good_for_meal")
store_table(good_for_meal_business_relation_table, "good_for_meal_business_relation")
store_table(dietary_restrictions_table, "dietary_restrictions")
store_table(dietary_restrictions_business_relation_table, "dietary_restrictions_business_relation")
store_table(day_table, "day")
store_table(schedule_table, "schedule")